test/connectors/httpParsers/StandardDownstreamHttpParserSpec.scala

```diff
@@ -0,0 +1,164 @@
+
+  val httpReads: HttpReads[DownstreamOutcome[Unit]] = implicitly
+
+  val data                            = "someData"
+  val downstreamExpectedJson: JsValue = Json.obj("data" -> data)
+
+  val downstreamModel: SomeModel                     = SomeModel(data)
+  val downstreamResponse: ResponseWrapper[SomeModel] = ResponseWrapper(correlationId, downstreamModel)
+
+  "The generic HTTP parser" when {
+    "no status code is specified" must {
+      val httpReads: HttpReads[DownstreamOutcome[SomeModel]] = implicitly
+
+      "return a Right downstream response containing the model object if the response json corresponds to a model object" in {
+        val httpResponse = HttpResponse(OK, downstreamExpectedJson, Map("CorrelationId" -> Seq(correlationId)))
+
+        httpReads.read(method, url, httpResponse) shouldBe Right(downstreamResponse)
+      }
+
+      "return an outbound error if a model object cannot be read from the response json" in {
+        val badFieldTypeJson: JsValue = Json.obj("incomeSourceId" -> 1234, "incomeSourceName" -> 1234)
+        val httpResponse              = HttpResponse(OK, badFieldTypeJson, Map("CorrelationId" -> Seq(correlationId)))
+        val expected                  = ResponseWrapper(correlationId, InternalError)
+
+        httpReads.read(method, url, httpResponse) shouldBe Left(expected)
+      }
+
+      handleErrorsCorrectly(httpReads)
+      handleInternalErrorsCorrectly(httpReads)
+      handleUnexpectedResponse(httpReads)
+    }
+  }
+
+  "The generic HTTP parser for empty response" when {
+    "no status code is specified" must {
+      val httpReads: HttpReads[DownstreamOutcome[Unit]] = implicitly
+
+      "receiving a 204 response" should {
+        "return a Right DownstreamResponse with the correct correlationId and no responseData" in {
+          val httpResponse = HttpResponse(NO_CONTENT, "", Map("CorrelationId" -> Seq(correlationId)))
+
+          httpReads.read(method, url, httpResponse) shouldBe Right(ResponseWrapper(correlationId, ()))
+        }
+      }
+
+      handleErrorsCorrectly(httpReads)
+      handleInternalErrorsCorrectly(httpReads)
+      handleUnexpectedResponse(httpReads)
+    }
+
+    "a success code is specified" must {
+      implicit val successCode: SuccessCode             = SuccessCode(PARTIAL_CONTENT)
+      val httpReads: HttpReads[DownstreamOutcome[Unit]] = implicitly
+
+      "use that status code for success" in {
+        val httpResponse = HttpResponse(PARTIAL_CONTENT, "", Map("CorrelationId" -> Seq(correlationId)))
+
+        httpReads.read(method, url, httpResponse) shouldBe Right(ResponseWrapper(correlationId, ()))
+      }
+    }
+  }
+
+  val singleErrorJson: JsValue = Json.parse(
+    """
+      |{
+      |   "code": "CODE",
+      |   "reason": "MESSAGE"
+      |}
+    """.stripMargin
+  )
+
+  val malformedErrorJson: JsValue = Json.parse(
+    """
+      |{
+      |   "coed": "CODE",
+      |   "resaon": "MESSAGE"
+      |}
+    """.stripMargin
+  )
```

Comment:
Maybe we can remove the `malformedErrorJson` as the tests that use it don't really rely on it being malformed (because the code does not look at the JSON body: it can be anything).